### PR TITLE
8255047: Add HotSpot UseDebuggerErgo flags

### PR DESCRIPTION
**Pull request by stefank**
Some debuggers don't work well with many threads, and/or incompletely restricts the number of used CPUs to one.

This flag is intended as a catch-all for HotSpot developers (not available in product builds) to allow us to more easily use those debuggers.

Currently, the proposal is to let the flag fix a few things:
1) Turn down the number of JVM threads
2) Turn off NUMA
3) Force processor_id() to return 0 instead of values above processor_count()

(1) is purely ergonomics: gdb, rr, valgrind is faster and seems to work much better with fewer threads. The values would still be overridable by devs.

(2) and (3) deals with the fact that some debuggers change the reported processor count, but don't change the processor ids returned by sched_getcpu. This causes problems for ZGC and NUMA, that both assumes that they can rely on os::processor_id() < os::processor_count().

The current proposed flag name is -XX:+LimitedCPUsDebugging. I'm not entirely happy with that name, but I been able to find a better name.

An alternative to having one flag, is to split this into two flags, and maybe that would solve the naming problem. However, the usability aspects will be worse.

If we can't find a suitable name, I rather introduce a flag called:
-XX:DebuggerWorkarounds or -XX:DebuggerWorkaround1 

Any suggestions / opinions? I really do want to at least fix the (2, 3) problem, because I keep having to add this to every single branch I'm working on.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Testing

|     | Linux x32 | Linux x64 | Windows x64 | macOS x64 |
| --- | ----- | ----- | ----- | ----- |
| Build | ✔️ (1/1 passed) | ✔️ (5/5 passed) | ✔️ (2/2 passed) | ✔️ (2/2 passed) |
| Test (tier1) |    |  ✔️ (9/9 passed) | ❌ (1/9 failed) | ✔️ (9/9 passed) |

**Failed test task**
- [Windows x64 (hs/tier1 compiler)](https://github.com/stefank/jdk/runs/1286739854)

### Issue
 * [JDK-8255047](https://bugs.openjdk.java.net/browse/JDK-8255047): Add HotSpot UseDebuggerErgo flags

### Reviewers
 * [Daniel D. Daugherty](https://openjdk.java.net/census#dcubed) (@dcubed-ojdk - **Reviewer**)
 * [David Holmes](https://openjdk.java.net/census#dholmes) (@dholmes-ora - **Reviewer**)

### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/763/head:pull/763`
`$ git checkout pull/763`

### CODE REVIEW

**Comment by dcubed-ojdk** [comment]
Perhaps: `UseDebuggerErgo` for the main option name with
`UseDebuggerErgo1` ... `UseDebuggerErgoN` for the suboptions
where `UseDebuggerErgo` enables all the numbered
`UseDebuggerErgo<X>` options in one go.

**Comment by stefank** [comment]
@dcubed-ojdk Thanks for the suggestion. I like it.

**Comment by stefank** [comment]
I've updated the patch with the suggestion from Dan. The flags now work as follows:
-XX:+UseDebuggerErgo turns on all ergonomics / workarounds
-XX:-/+UseDebuggerErgo1 fixes processor id vs processor count problems
-XX:-/+UseDebuggerErgo2 limits the number of JVM threads and turns of NUMA

File: src/hotspot/os/linux/os_linux.cpp [hunk 1]
```diff
@@ -4762,7 +4772,16 @@ int os::active_processor_count() {
...
+
+#ifndef PRODUCT
+  if (UseDebuggerErgo1 && id >= _processor_count) {
+    // Some debuggers limit the processor count without limiting
+    // the returned processor ids. Fake the processor id.
+    return 0;
+  }
+#endif
+
+  assert(id >= 0 && id < _processor_count, "Invalid processor id [%d]", id);
```
Comment (dcubed-ojdk) [inline src/hotspot/os/linux/os_linux.cpp:4784]: Thanks for adding the bad id value to the assert() output.

**Review by dcubed-ojdk — APPROVED** [review]
Thumbs up.

**Review by dholmes-ora — APPROVED** [review]
LGTM!

Thanks,
David

**Comment by stefank** [comment]
Thanks @dcubed-ojdk and @dholmes-ora!

**Pull request closed without merging**